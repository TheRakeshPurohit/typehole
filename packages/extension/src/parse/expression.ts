import * as ts from "typescript";

export function isValidSelection(selectedNode: ts.Node) {
  const kind = ts.SyntaxKind[selectedNode.kind];
  return (
    kind.includes("Expression") ||
    kind.includes("Literal") ||
    kind.includes("Keyword")
  );
}
