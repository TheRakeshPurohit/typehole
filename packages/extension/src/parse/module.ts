import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

export function findTypeHoleImports(ast: ts.Node) {
  return tsquery
    .query(ast, "ImportDeclaration > StringLiteral[text='typehole']")
    .map((s) => s.parent);
}

export function resolveImportPath(
  projectRoot: string,
  moduleName: string,
  containingFile: string
) {
  const configFileName = ts.findConfigFile(
    projectRoot,
    ts.sys.fileExists,
    "tsconfig.json"
  );

  if (!configFileName) {
    return null;
  }

  const configFile = ts.readConfigFile(configFileName, ts.sys.readFile);

  const compilerOptions = ts.parseJsonConfigFileContent(
    configFile.config,
    ts.sys,
    "./"
  );

  function fileExists(fileName: string): boolean {
    return ts.sys.fileExists(fileName);
  }

  function readFile(fileName: string): string | undefined {
    return ts.sys.readFile(fileName);
  }

  const result = ts.resolveModuleName(
    moduleName,
    containingFile,
    compilerOptions.options,
    {
      fileExists,
      readFile,
    }
  );

  return result.resolvedModule!.resolvedFileName;
}

export function findLastImport(ast: ts.Node) {
  const imports = tsquery.query(ast, "ImportDeclaration");
  return imports[imports.length - 1];
}

export function getNodeEndPosition(node: ts.Node) {
  return node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd());
}
export function getNodeStartPosition(node: ts.Node) {
  return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart());
}

export function getTypeHoleImport() {
  const clause = ts.factory.createImportClause(
    false,
    ts.factory.createIdentifier("typehole"),
    undefined
  );
  return printAST(
    ts.factory.createImportDeclaration(
      undefined,
      undefined,
      clause,
      ts.factory.createStringLiteral("typehole")
    )
  );
}

export function printAST(ast: ts.Node, sourceFile?: ts.SourceFile) {
  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

  return printer.printNode(
    ts.EmitHint.Unspecified,
    ast,
    sourceFile || ast.getSourceFile()
  );
}

export function findTypeholes(ast: ts.Node | string): ts.CallExpression[] {
  const holes = tsquery.query(
    ast,
    `PropertyAccessExpression > Identifier[name="typehole"]`
  );

  return holes
    .map((n) => n.parent.parent)
    .filter(ts.isCallExpression)
    .sort((a, b) => {
      const keyA = (a.expression as ts.PropertyAccessExpression).name.getText();
      const keyB = (b.expression as ts.PropertyAccessExpression).name.getText();
      return keyA.localeCompare(keyB);
    });
}

export function getAST(source: string) {
  return tsquery.ast(source, "file.ts", ts.ScriptKind.TSX);
}

export function getParentOnRootLevel(node: ts.Node): ts.Node {
  if (ts.isSourceFile(node.parent)) {
    return node;
  }
  return getParentOnRootLevel(node.parent);
}
export function someParentIs(
  node: ts.Node,
  test: (node: ts.Node) => boolean
): boolean {
  if (!node.parent) {
    return false;
  }
  if (test(node.parent)) {
    return true;
  }
  return someParentIs(node.parent, test);
}

export function getParentWithType<T = ts.Node>(
  node: ts.Node,
  kind: ts.SyntaxKind
): T | null {
  if (!node.parent) {
    return null;
  }
  if (node.kind === kind) {
    return node as unknown as T;
  }
  return getParentWithType(node.parent, kind);
}
