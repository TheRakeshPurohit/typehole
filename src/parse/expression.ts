import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

export function isExpression(text: string) {
  const ast = tsquery.ast(text);
  const selectedNodes = ast.getChildAt(0).getChildren();

  const onlyExpressionAndEndOfFile =
    selectedNodes.length === 1 && ts.isExpressionStatement(selectedNodes[0]);

  if (onlyExpressionAndEndOfFile) {
    return true;
  }
  return false;
}
