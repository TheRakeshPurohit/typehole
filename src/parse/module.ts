import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

export function findTypeHoleImport(ast: ts.Node) {
  return tsquery.query(
    ast,
    "ImportDeclaration > StringLiteral[text='typehole']"
  );
}
export function findLastImport(ast: ts.Node) {
  const imports = tsquery.query(ast, "ImportDeclaration");
  return imports[imports.length - 1];
}

export function getNodeEndPosition(node: ts.Node) {
  return node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd());
}
export function getNodeStartPosition(node: ts.Node) {
  return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart());
}

export function getTypeHoleImport() {
  const clause = ts.factory.createImportClause(
    false,
    ts.factory.createIdentifier("typehole"),
    undefined
  );
  return printAST(
    ts.factory.createImportDeclaration(
      undefined,
      undefined,
      clause,
      ts.factory.createStringLiteral("typehole")
    )
  );
}

export function getTypeHoleFactoryCall(id: number) {
  return `const t${id === 0 ? "" : id} = typehole()`;
}

export function printAST(ast: ts.Node, sourceFile?: ts.SourceFile) {
  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });

  return printer.printNode(
    ts.EmitHint.Unspecified,
    ast,
    sourceFile || ast.getSourceFile()
  );
}

export function findTypeholes(ast: ts.Node | string): ts.CallExpression[] {
  const holes = tsquery.query(
    ast,
    `PropertyAccessExpression > Identifier[name="typehole"]`
  );

  return holes
    .map((n) => n.parent.parent)
    .filter(ts.isCallExpression)
    .sort((a, b) => {
      const keyA = (a.expression as ts.PropertyAccessExpression).name.getText();
      const keyB = (b.expression as ts.PropertyAccessExpression).name.getText();
      return keyA.localeCompare(keyB);
    });
}

export function getAST(source: string) {
  return tsquery.ast(source, "file.ts", ts.ScriptKind.TSX);
}

export function getParentOnRootLevel(node: ts.Node): ts.Node {
  if (ts.isSourceFile(node.parent)) {
    return node;
  }
  return getParentOnRootLevel(node.parent);
}
export function someParentIs(
  node: ts.Node,
  test: (node: ts.Node) => boolean
): boolean {
  if (!node.parent) {
    return false;
  }
  if (test(node.parent)) {
    return true;
  }
  return someParentIs(node.parent, test);
}
