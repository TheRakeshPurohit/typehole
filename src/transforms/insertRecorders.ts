import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

export function insertRecorders(source: string) {
  const ast = tsquery.ast(source);
  const sourceFile = ast.getSourceFile();

  const magicMarkers = tsquery.query(
    ast,
    "TypeReference > Identifier[name='AutoDiscover']"
  );

  if (magicMarkers.length === 0) {
    return source;
  }

  const variableDeclarations = magicMarkers.map((m) => m.parent.parent);
  const markerStarts = magicMarkers.map((m) => m.parent);

  function transformer(ctx: ts.TransformationContext) {
    return function transform(source: ts.SourceFile) {
      function visitor(node: ts.Node): ts.Node | ts.Node[] {
        if (variableDeclarations.includes(node.parent)) {
          const start = node;
          const siblings = start.parent.getChildren();
          const isLast = siblings.indexOf(node) === siblings.length - 1;
          if (!isLast) {
            return ts.visitEachChild(node, visitor, ctx);
          }
          if (!siblings.some((s) => markerStarts.includes(s))) {
            return ts.visitEachChild(node, visitor, ctx);
          }

          return ts.factory.createCallExpression(
            ts.factory.createIdentifier("t"),
            undefined,
            [node as ts.Expression]
          );
        }
        return ts.visitEachChild(node, visitor, ctx);
      }
      return ts.visitNode(source, visitor);
    };
  }

  const transformed = ts.transform(ast, [transformer]);

  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
  const result = printer.printNode(
    ts.EmitHint.Unspecified,
    transformed.transformed[0],
    sourceFile
  );
  return result;
}
