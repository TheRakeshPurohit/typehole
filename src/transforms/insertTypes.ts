import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";
import {
  findTypeholeFactories,
  findTypeholes,
  getNodeEndPosition,
  printAST,
  getNodeStartPosition,
} from "../parse/module";

function lineCharacterPositionInText(
  lineChar: ts.LineAndCharacter,
  text: string
) {
  const rows = text.split("\n");

  const allLines = rows.slice(0, lineChar.line + 1);
  allLines[allLines.length - 1] = allLines[allLines.length - 1].substr(
    0,
    lineChar.character
  );

  return allLines.join("\n").length;
}

function insertBefore(source: string, node: ts.Node, insert: string) {
  const start = getNodeStartPosition(node);
  const end = getNodeEndPosition(node);

  return (
    source.substr(0, lineCharacterPositionInText(start, source)) +
    insert +
    source.substr(lineCharacterPositionInText(start, source))
  );
}

function replace(source: string, node: ts.Node, replacement: string) {
  const start = getNodeStartPosition(node);
  const end = getNodeEndPosition(node);

  return (
    source.substr(0, lineCharacterPositionInText(start, source)) +
    replacement +
    source.substr(lineCharacterPositionInText(end, source))
  );
}

export function insertTypes(id: number, source: string, types: string) {
  const ast = tsquery.ast(source);
  const typeId = `AutoDiscover${id === 0 ? "" : id}`;
  const typeString = types.trim().replace("AutoDiscover", typeId);

  const existingTypeDeclaration = tsquery.query(
    ast,
    `InterfaceDeclaration > Identifier[name='${typeId}']`
  );
  let newSource = source;
  if (existingTypeDeclaration.length > 0) {
    newSource = replace(source, existingTypeDeclaration[0].parent, typeString);
  } else {
    const factory = findTypeholeFactories(ast)[id];

    newSource = insertBefore(source, factory, `${typeString}\n`);
  }
  const newAST = tsquery.ast(newSource);
  const holes = findTypeholes(newAST);
  const hole = holes[id];

  if (!hole) {
    console.error("Cannot find typehole with an id", id);
    return newSource;
  }

  if (ts.isVariableDeclaration(hole.parent)) {
    const existingType = tsquery.query(
      hole.parent,
      `TypeReference > Identifier[name='${typeId}']`
    );
    if (existingType.length === 0) {
      const newVariableDeclaration = ts.factory.createVariableDeclaration(
        hole.parent.name,
        hole.parent.exclamationToken,
        ts.factory.createTypeReferenceNode(typeId),
        hole.parent.initializer
      );

      newSource = replace(
        newSource,
        hole.parent,
        printAST(newVariableDeclaration, newAST.getSourceFile())
      );
    }
  }

  return newSource;
}
