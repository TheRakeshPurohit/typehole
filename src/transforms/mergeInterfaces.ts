import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

function reduceNodes<T>(
  node: ts.Node,
  op: (memo: T, node: ts.Node) => T,
  defaultValue: T
): T {
  const children = node.getChildren();

  return children.reduce((memo, child) => {
    const newVal = op(memo, child);
    return reduceNodes(child, op, newVal);
  }, defaultValue);
}

export function mergeInterfaces(source: string) {
  const ast = tsquery.ast(source);
  const sourceFile = ast.getSourceFile();

  const getParentInterfaceName = (node: ts.Node): ts.Identifier =>
    ts.isInterfaceDeclaration(node)
      ? node.name
      : getParentInterfaceName(node.parent);

  const knownInterfaces = reduceNodes<Record<string, ts.InterfaceDeclaration>>(
    ast.getChildAt(0),
    (memo, node) => {
      if (ts.isInterfaceDeclaration(node)) {
        return { ...memo, [node.name.text]: node };
      }
      return memo;
    },
    {}
  );

  const dependencies = reduceNodes<Record<string, string[]>>(
    ast.getChildAt(0),
    (memo, node) => {
      if (ts.isTypeReferenceNode(node)) {
        const interfaceName = getParentInterfaceName(node).text;

        return {
          ...memo,
          [interfaceName]: memo[interfaceName]
            ? memo[interfaceName].concat(node.getText())
            : [node.getText()],
        };
      }
      return memo;
    },
    Object.fromEntries(Object.keys(knownInterfaces).map((key) => [key, []]))
  );

  const someoneDependingOn = Object.values(dependencies).flat();
  const roots = Object.keys(dependencies).filter(
    (type) => !someoneDependingOn.includes(type)
  );

  function interfaceToTypeLiteral(
    node: ts.InterfaceDeclaration | ts.TypeLiteralNode
  ): ts.TypeLiteralNode {
    return ts.factory.createTypeLiteralNode(
      node.members.map(expandTypeElement)
    );
  }

  function expandTypeElement(member: ts.TypeElement) {
    const type = (member as any).type;
    if (ts.isTypeReferenceNode(type)) {
      if (knownInterfaces[type.typeName.getText()]) {
        return ts.factory.createPropertySignature(
          undefined,
          member.name as ts.PropertyName,
          member.questionToken,
          interfaceToTypeLiteral(knownInterfaces[type.typeName.getText()])
        );
      }
    }
    if (ts.isTypeLiteralNode(type)) {
      return ts.factory.createPropertySignature(
        undefined,
        member.name!,
        member.questionToken,
        interfaceToTypeLiteral(type)
      );
    }
    if (ts.isArrayTypeNode(type) && ts.isTypeReferenceNode(type.elementType)) {
      if (knownInterfaces[type.elementType.typeName.getText()]) {
        return ts.factory.createPropertySignature(
          undefined,
          member.name!,
          member.questionToken,
          interfaceToTypeLiteral(
            knownInterfaces[type.elementType.typeName.getText()]
          )
        );
      }
    }
    if (
      ts.isArrayTypeNode(type) &&
      ts.isParenthesizedTypeNode(type.elementType)
    ) {
      if (ts.isUnionTypeNode(type.elementType.type)) {
        const types = type.elementType.type.types.filter(
          (t) => knownInterfaces[t.getText()]
        );

        return ts.factory.createPropertySignature(
          undefined,
          member.name!,
          member.questionToken,
          ts.factory.createArrayTypeNode(
            ts.factory.createParenthesizedType(
              ts.factory.createUnionTypeNode(
                types.map((t) =>
                  interfaceToTypeLiteral(knownInterfaces[t.getText()])
                )
              )
            )
          )
        );
      }
    }

    return member;
  }
  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
  const mergedInterfaces = roots
    .map((root) => knownInterfaces[root])
    .map((node: ts.InterfaceDeclaration) => {
      return ts.factory.createInterfaceDeclaration(
        undefined,
        undefined,
        node.name,
        undefined,
        undefined,
        node.members.map(expandTypeElement)
      );
    })
    .map((transformed) =>
      printer.printNode(ts.EmitHint.Unspecified, transformed, sourceFile)
    );

  return mergedInterfaces.join("\n");
}
