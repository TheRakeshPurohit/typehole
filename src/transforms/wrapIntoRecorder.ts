import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

export function wrapIntoRecorder(id: number, expressionCode: string) {
  const ast = tsquery.ast(expressionCode);
  const sourceFile = ast.getSourceFile();
  const expression = ast.getChildAt(0).getChildAt(0).getChildAt(0);

  const wrapped = ts.factory.createCallExpression(
    ts.factory.createIdentifier(`typehole.t${id === 0 ? "" : id}`),
    undefined,
    [expression as ts.Expression]
  );

  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
  const result = printer.printNode(
    ts.EmitHint.Unspecified,
    wrapped,
    sourceFile
  );
  return result;
}
