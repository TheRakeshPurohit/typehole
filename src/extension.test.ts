import { tsquery } from "@phenomnomnominal/tsquery";
import * as assert from "assert";

import { isExpression } from "./parse/expression";
import { findTypeholes, getAST } from "./parse/module";

import { getAllDependencyTypeDeclarations } from "./transforms/insertTypes";

test("finds selected expression", () => {
  const actual = isExpression(`
  tsquery.query(
    ast,
    "InterfaceDeclaration > Identifier[name='AutoDiscover']"
  )
  `);

  assert.deepStrictEqual(actual, true);
});
test("finds selected expression", () => {
  const actual = isExpression(`
  (await axios.post(
    "https://www.etuovi.com/api/v2/announcements/search/listpage",
    params
  )
).data
  `);

  assert.deepStrictEqual(actual, true);
});

test("returns null on non-expression selection", () => {
  const actual = isExpression(`
  if (!siblings.some((s) => markerStarts.includes(s))) {
    return ts.visitEachChild(node, visitor, ctx);
  }
  `);
  assert.strictEqual(actual, false);
});

test("returns null on non-expression selection", () => {
  const actual = isExpression(`
  .some((s) => markerStarts.includes(s))) {
    return ts.visitEachChild(node, visitor, ctx);
  }
  `);
  assert.strictEqual(actual, false);
});

test("finds all typewholes from source", () => {
  const actual = findTypeholes(`
  import { VercelRequest, VercelResponse } from "@vercel/node";
import axios from "axios";
import typehole from "typehole";

const t3 = typehole();

export default async (request: VercelRequest, response: VercelResponse) => {
  const xsrf = await getXSRF();
  const res = t3(
    (
      await axios.post(
        "https://www.etuovi.com/api/v2/announcements/search/listpage",
        params
      )
    ).data
  );

  return response.status(200).send(res.announcements);
};

  `);
  assert.strictEqual(actual.length, 1);
});

test("Finds all dependency type nodes from an AST", () => {
  const ast = getAST(`type Something = {
    a: B;
  };

  type B = {
    moi: C | D;
  };

  type C = 2;

  type D = 3;`);
  const actual = getAllDependencyTypeDeclarations(
    tsquery.query(ast, ':declaration > Identifier[name="Something"]')[0].parent
  );

  expect(actual.map((n) => n.getText())).toMatchSnapshot();
});
