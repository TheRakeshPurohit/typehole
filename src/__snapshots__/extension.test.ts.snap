// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Finds all dependency type nodes from an AST 1`] = `
Array [
  "type Something = {
    a: B;
  };",
  "type B = {
    moi: C | D;
  };",
  "type C = 2;",
  "type D = 3;",
]
`;
